### PR TITLE
remove name_to_handle_at(2) from filtered syscalls

### LINKED ISSUE #45518: name_to_handle_at is gated by CAP_SYS_ADMIN and not CAP_DAC_READ_SEARCH as documented
### Description

The docs state:
> https://docs.docker.com/engine/security/seccomp/
>  name_to_handle_at | Sister syscall to open_by_handle_at. Already gated by CAP_DAC_READ_SEARCH.

But the profile does not include this syscall in the CAP_DAC_READ_SEARCH, forcing use of CAP_SYS_ADMIN if you need this syscall.

https://github.com/moby/moby/blob/e58c267d66cd9c8fc89c582047b22158c9d44db8/profiles/seccomp/default.json#L580-L588

Consequently the syscall fails with EPERM (errno -1).

### Reproduce

N/A

### Expected behavior

It should be possible to enable this syscall with `docker run --cap-add CAP_DAC_READ_SEARCH`.

### docker version

```bash
20.10.23 and e58c267d66cd9c8fc89c582047b22158c9d44db8.
```

### docker info

```bash
N/A
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by bartier**
closes #45518

Hi there, this is my first PR here, so please feel free to point me out if anything is wrong with this contribution. I looked the #45518 issue and I believe this could be a fix.

**- What I did**
Removed the function from the filtered syscalls as `name_to_handle_at(2)` is in fact innocuous and safe

**- How I did it**
@neersighted help at [comment](https://github.com/moby/moby/pull/45766#pullrequestreview-1493908145)

**- How to verify it**
N/A

**- Description for the changelog**

Remove name_to_handle_at(2) from filtered syscalls

**- A picture of a cute animal (not mandatory but encouraged)**
image

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks for contributing! I see nobody posted here yet, but @neersighted has been digging into history (as well as Linux kernel man-pages and source-code) to verify these changes. I think some further digging was still needed, as there were some (very subtle?) cases where this could cause issues (stay tuned 👍)

(Code changes themselves look good 👍, but it may need a rebase to make the `docker-py` tests pass (which had an issue that has been fixed on master), but we can ignore those for the time being)

**Review by neersighted — CHANGES_REQUESTED** [review]
Thanks for sending this patch!

Since this load-bearing (lots of useful software, e.g. systemd wants this syscall), (potentially) security critical, and has been in this state for a very long time, I did some extensive digging.

First, some background on what `name_to_handle_at(2)` and `open_by_handle_at(2)` actually do: this is more-or-less a two-stage `openat(2)`, as [described in the manpage](https://manpages.ubuntu.com/manpages/jammy/en/man2/open_by_handle_at.2.html).

`name_to_handle_at(2)` performs a filesystem lookup, but instead of opening a file, it copies a (opaque) `struct file_handle` to userspace. That handle contains a string, which is [looked up](https://elixir.bootlin.com/linux/v5.19.17/source/fs/fhandle.c#L49) in the kernel as part of the system call.

`open_by_handle_at(2)` can then be used to open the path described by that file handle (which is [dereferenced](https://elixir.bootlin.com/linux/v5.19.17/source/fs/fhandle.c#L152) similarly to how it was saved), without doing a filesystem lookup a second time. Note that this does not defend against TOCTOU issues (e.g. the filesystem changing between calls); it merely allows one to pay the cost of resolving a path in the filesystem once, and amortize that cost over many usages.

This appears to have been added to the kernel to better support userspace network filesystem servers (which open and close **many** files and thus need all the performance shortcuts they can get), and the kernel implementation actually builds on the infrastructure of the `exportfs` subsystem used for the in-kernel NFS implementation.

The curious among us might find the following parts of the kernel source interesting:
* [fs/fhandle:handle_to_path](https://elixir.bootlin.com/linux/v5.19.17/source/fs/fhandle.c#L167)
* [fs/namei:do_filp_open](https://elixir.bootlin.com/linux/v5.19.17/source/fs/namei.c#L3676)
* [fs/namei:do_file_open_root](https://elixir.bootlin.com/linux/v5.19.17/source/fs/namei.c#L3693)
* [fs/namei:do_open](https://elixir.bootlin.com/linux/v5.19.17/source/fs/namei.c#L3478)

Nonetheless, there are some differences from `openat(2)`: `open_by_handle_at(2)` requires [`capabilities(7)`](https://manpages.ubuntu.com/manpages/jammy/en/man7/capabilities.7.html); specifically, `CAP_DAC_READ_SEARCH`, but `openat(2)` requires no privilege. Why is that?

First, let's establish what does **not** make `open_by_handle_at(2)` dangerous:
* `open_by_handle_at(2)` does pass a opaque and kernel-provided data structure back in, over the kernel/userspace boundary. However, the kernel is quite resilient to potentially hostile inputs, and in this specific case, `struct file_handle` is just structured data and is safely deserialized -- the kernel never tries to use what it gets back from userspace as a normal internal data structure.
* Permissions are checked (the thing that the related `CAP_DAC_OVERRIDE` capability allows bypassing) during `open_by_handle_at(2)`, in the common [`do_open`](https://elixir.bootlin.com/linux/v5.19.7/source/fs/namei.c#L3520) used by the entire `open(2)` family of system calls, so there is no risk of a stored `struct file_handle` being used to bypass (Unix) permissions.

If there is no risk of attacking the kernel, or bypassing DAC, what is the risk of this system call then? Well, if the user can construct arbitrary `struct file_handle`s (hint: they can) and pass them to the kernel, there is the risk they manage to forge a handle and open a path that they otherwise could not access. Note that 'access' is not in the Unix permissions sense -- I use 'access' here to mean "open a `dentry` that cannot otherwise be resolved using the visible filesystem." 

It's notable that there are several situations where filesystem access is not limited by permissions (e.g. a program has root), but instead by filesystem visibility:
* `chroot(2)`
* mount namespaces (including `pivot_root(2)`), which are **used in containers** (more on this later)

Indeed, if we look at the history of the patch that introduced these system calls:
* The [initial version of the patch](https://lore.kernel.org/all/1266558149-11460-3-git-send-email-aneesh.kumar@linux.vnet.ibm.com/) required `CAP_SYS_ADMIN` (infamously, ['the new root'](https://lwn.net/Articles/486306/)).
  - The risk of `struct file_handle` being forged was brought up in [one of the earliest reviews](https://lore.kernel.org/all/AD5A6061-9B7B-4B29-A458-822E781C2588@sun.com/).
* The [next version of the patch](https://lore.kernel.org/all/1268932144-14105-1-git-send-email-aneesh.kumar@linux.vnet.ibm.com/) moved to `CAP_DAC_OVERRIDE`.
  - Presumably the logic here was that forging a handle was similar to 'a process that is trusted to ignore Unix permissions'.
* [Subsequent versions](https://lore.kernel.org/all/1275581657-7232-1-git-send-email-aneesh.kumar@linux.vnet.ibm.com/) decreased the capability to `CAP_DAC_READ_SEARCH`.
  - As the only risk here is resolving a path that you otherwise could not, 'read' + 'search' (resolve path) are presumably a more granular fit.
  - The patch does note that `CAP_DAC_SEARCH` would be preferred if it existed, as it would more accurately reflect the threat model.
* [The final, merged version of the patch](https://lore.kernel.org/all/1296328099-23263-1-git-send-email-aneesh.kumar@linux.vnet.ibm.com/) made no changes to the permissions/capabilities model.

Given the above context, it is very clear that `name_to_handle_at(2)` is a safe operation (all it does is return a value that is useless without its sister syscall), and `open_by_handle_at(2)` is the dangerous component. So why is `name_to_handle_at(2)` more restricted?

Before we address that point directly, it helps to think about how Linux containers work (and how they isolate filesystem access). As noted above, we use mount [namespaces(7)](https://manpages.ubuntu.com/manpages/jammy/man7/namespaces.7.html) as part of the containerization. In fact, mount namespaces are one of the most critical forms of isolation making up a container; you can have root inside a container (and thus are allowed to modify any file), but you can't use that root access to affect files outside the container if you can't see the filesystem.

This is in fact false if you can abuse `open_by_handle_at(2)` to get access to a path you otherwise could not see in your mount namespace; this is the basis of one of the first container escapes, 'shocker.'

Others have already done an excellent job at explaining the exploit; these links should give more background if it is interesting to you:
* http://stealth.openwall.net/xSports/shocker.c
* https://lxc-users.linuxcontainers.narkive.com/9kzuZR09/container-escape-through-open-by-handle-at-shocker-exploit
* https://medium.com/@fun_cuddles/docker-breakout-exploit-analysis-a274fff0e6b3

This was ultimately solved in Docker by [dropping capabilities by default](https://github.com/moby/moby/commit/9d6875d19d3926faf6287487234ad0b2f1310e9d), beyond a narrow list of innocuous ones (e.g. `CAP_MKNOD`).

Now we understand this family of system calls, and their risk to containers. However, we still haven't puzzled out why `name_to_handle_at(2)` was gated behind `CAP_SYS_ADMIN`; to begin understanding that, let's look at the evolution of the default seccomp profile:
* [The first version](https://github.com/jessfraz/docker/commit/947293a28084cb5ee2e10e4d128c6e2b9d9da89d#diff-617cddaebc1111c35f3ea1b55908653720ec669df60f508bbe182f83bc794a6cR185-R189) of the profile blocked `open_by_handle_at(2)` as it was implicated in shocker.
  - This appears to be something of a 'why not?' situation; the actual escape is solved by dropping the capability, but it doesn't hurt to filter with seccomp as well...
* [A subsequent change](https://github.com/moby/moby/commit/17735c3c98056006b40834d7426d8d90afae5a2c) was to move to an allowlist instead of a denylist; the explicit reference to `open_by_handle_at(2)` was dropped.
* [A later, major refactor](https://github.com/moby/moby/commit/a01c4dc8f85827f32d88522e5153dddc02f11806#diff-6c0d906dbef148d2060ed71a7461907e5601fea78866e4183835c60e5d2ff01aR1627-R1639) redesigned the default seccomp profile to work like it does today: `--cap-add` implies changes to the seccomp profile (and the profile is once again a denylist). At this time, both `open_by_handle_at(2)` and `name_to_handle_at(2)` were gated on `CAP_DAC_READ_SEARCH`.
  - This appears to be a case of 'guilty by association', i.e. both syscalls share a manpage, and one is behind a capability, so why not block them both?
  - Fun fact: the docs used to mistakenly assert this was [gated by `CAP_SYS_NICE`](https://github.com/moby/moby/blob/e35f5a481ac5c982c4f97be5dc41206e21e19007/docs/security/seccomp.md?plain=1#L117).
* [Finally, `name_to_handle_at(2)` was moved to `CAP_SYS_ADMIN`](https://github.com/moby/moby/commit/c1ca124682a90f3306b34ad104ba80e413f7bf88) because systemd required it, and did not actually require `open_by_handle_at(2)`.
  - It appears that no one reexamined **why** this was behind a capability, and assumed that it 'must' be because it was in some way dangerous.

---

In summary: this was incorrectly moved from `CAP_DAC_READ_SEARCH` to `CAP_SYS_ADMIN` (a more restrictive capability) in order to make it easier to run systemd (since systemd implied `--cap-add SYS_ADMIN`, and it made the command line shorter). However, `name_to_handle_at(2)` is in fact innocuous and safe, and **I assert it should not be filtered in the default seccomp profile at all.**

I suggest altering this PR to remove it from the list of filtered syscalls, and creating a new PR to adjust the docs.

**Comment by neersighted** [comment]
Also, after the above change is made and accepted, this PR will need backports to all branches (I can do this). In addition, we'll need to adjust the [profile in containerd](https://github.com/thaJeztah/containerd/blob/bbb8d3470483c8054fd435d815ba699f94350928/contrib/seccomp/seccomp_default.go#L572); however, I am happy to do that work.

**Comment by tianon** [comment]
Fantastic summary, and well done on the sleuthing, @neersighted :bow: :+1:

**Comment by thaJeztah** [comment]
@bartier do you need help / details on making the suggested changes? (I'm sure @neersighted is also happy to update this PR if needed)

**Comment by bartier** [comment]
Hi @neersighted , thanks for the depth explanation. Also, I have created the PR https://github.com/docker/docs/pull/17623 to match the changes.

Please let me know if I am missing something but I am happy to help.

**Comment by neersighted** [comment]
Thanks! The change overall looks good to me now; could we update the PR title and description for the new direction, and maybe rephrase the commit message to be a little more obvious (e.g. `remove name_to_handle_at(2) from filtered syscalls`)?

Otherwise, LGTM :+1:

**Comment by bartier** [comment]
@neersighted I've squashed to 1 commit and changes the PR title. I wonder if it is required to squash the commits as we can merge and squash in the GitHub UI.

**Comment by corhere** [comment]
> I wonder if it is required to squash the commits as we can merge and squash in the GitHub UI.

We have disabled the "Squash and merge" option on this repository. We want the merge commit to be separate from the PR commit(s), which doesn't happen when GitHub squash-merges a PR.

**Comment by neersighted** [comment]
Still needs the commit message updated, but otherwise LGTM

**Comment by bartier** [comment]
@neersighted Changed to match the PR title

**Review by neersighted — APPROVED** [review]
LGTM! Thank you for your patience!

**Review by tianon — APPROVED** [review]

**Comment by neersighted** [comment]
Backports at:
- https://github.com/moby/moby/pull/45833
- https://github.com/moby/moby/pull/45834
- https://github.com/moby/moby/pull/45835

PTAL @bartier

**Comment by neersighted** [comment]
containerd PRs at:
- https://github.com/containerd/containerd/pull/8752
- https://github.com/containerd/containerd/pull/8753
- https://github.com/containerd/containerd/pull/8754

File: profiles/seccomp/default.json [hunk 1]
```diff
@@ -601,7 +601,6 @@
 				"mount",
 				"mount_setattr",
 				"move_mount",
-				"name_to_handle_at",
```
Comment (thaJeztah) [inline profiles/seccomp/default.json:604]: @neersighted I noticed we remove this one from the list here, but don't add it to the allow-list, https://github.com/moby/moby/blob/a78c06e3f06e6f55518e1ba2bf9f723cf19d0a65/profiles/seccomp/default_linux.go#L401

Doesn't that mean that it now gets the default (`specs.ActErrno` /  `SCMP_ACT_ERRNO`)?
https://github.com/moby/moby/blob/a78c06e3f06e6f55518e1ba2bf9f723cf19d0a65/profiles/seccomp/default_linux.go#L818
